### DISCUSSION TITLE
ssd1306.py text orientation question

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#12425] [Hardware & Peripherals]
I've been using the ssd1306.py lib to run the oled display on some TTGO lora boards. In the init_display section there are a couple of lines that allow the display text  to be turned upside down depending on the orientation of the board:
SET_SEG_REMAP | 0x01 changes to SET_SEG_REMAP to flip the horizontal
and 
SET_COM_OUT_DIR | 0x08 changes to SET_COM_OUT_DIR to flip the vertical

Editing these 2 lines has worked great till today when I stumbled across a board where the text will not re-orient. After I got over the disbelief I had a close look at the board. It's maybe a year or two older than its brethren, stamped lilygo rather than TTGO like the others, the flexible pcb ribbon to the oled is a bit narrower and the red power led is before the on/off switch rather than after it as per the rest. 

I'm no expert on display drivers but if it runs the oled OK surely the text re-orientation edit should work? Anybody seen similar recalcitrance before? More importantly, anybody know a work around?

**Comment by jimmo** [comment]
> I'm no expert on display drivers but if it runs the oled OK surely the text re-orientation edit should work?

This isn't necessarily true. It's possible that the controller on the OLED module doesn't implement all features.

**Comment by mcauser** [comment]
Rather than editing `init_display()`, there is a `rotate()` method:
https://github.com/micropython/micropython-lib/blob/master/micropython/drivers/display/ssd1306/ssd1306.py#L92C9-L92C15

```
init_display()
0xA1  # SET_SEG_REMAP
0xC8  # SET_COM_OUT_DIR

rotate(1) - same as init_display()
0xA1  # SET_SEG_REMAP
0xC8  # SET_COM_OUT_DIR

rotate(0)
0xA0  # SET_SEG_REMAP
0xC0  # SET_COM_OUT_DIR
```

Are you sure it's a SSD1306, or maybe one of the newer SSD1315 modules?
If it's the Lilygo display, I believe it's a SH1106 display module.
Can you attach photo(s)?

**Reply by kjm1102** [reply]
Something weird going on. I tried rotate(1) & rotate(0) & the text flips. I should be happy but my ssd1306 lib has no rotate functionality (I checked with a text search, the word 'rotate' is not present). How is this possible?

**Reply by robert-hh** [reply]
Did you get a copy of the driver linked by @mcauser ?

**Comment by uGinn-R** [comment]
Those displays may have different controllers. They can both work, but the functionality may differ.. So you have to refer to the manufacturer's documentation.

**Comment by kjm1102** [comment]
I had a look at it Rob to see what the rotate CMD does but I haven't loaded
it on the esp32 yet, so I don't understand why oled.rotate() is working
with my old lib.

**Reply by kjm1102** [reply]
OK I finally figured this out. It is the ssd1306.py built into upython that is being imported. So it doesn't matter what lib I have on the esp32, the import grabs the upython version that must include the rotate functionality. The penny finally dropped when I deleted the lib folder on the esp32 & the import ssd1306 cmd still worked. This also explains the problem that started this thread, the manual edits to flip the text were not working because that lib was not being used.

I'm fairly sure there didn't used to be an ssd1306.py in upython when I first started working with oled displays a year or two back but clearly there is one now.

**Reply by jimmo** [reply]
The following boards "freeze" ssd1306 by default:

```
LILYGO_TTGO_LORA32
LOLIN_S2_PICO
GARATRONIC_PYBSTICK26_F411
ESP8266_GENERIC
```

The first three are because the authors of those boards requested it, the final one is for historical reasons with the ESP8266 port.

**Reply by kjm1102** [reply]
Do u have a link to that lib? I like to have a copy of any lib I'm using so I can look up line numbers if it throws an error.

**Reply by jimmo** [reply]
It's the one in micropython-lib -- https://github.com/micropython/micropython-lib/tree/master/micropython/drivers/display/ssd1306
